blaze/expr/collections.py

```diff
@@ -435,22 +435,31 @@ def schema(self):
         """
         option = lambda dt: dt if isinstance(dt, Option) else Option(dt)
 
-        joined = [[name, dt] for name, dt in self.lhs.schema[0].parameters[0]
-                  if name in self.on_left]
+        on_left = self.on_left
+        on_right = self.on_right
+
+        right_params = self.rhs.schema[0].parameters[0]
+        joined = [
+            [name, extract_option(dt)
+             if isinstance(dt, Option) and
+             not isinstance(right_params[n], Option) else dt]
```

Comment:
yeah, I am changing this up a bit to use: https://github.com/blaze/datashape/pull/172